sql/sql_parse.cc

```diff
@@ -5153,9 +5153,41 @@ mysql_execute_command(THD *thd)
       my_ok(thd);
     break;
   case SQLCOM_BACKUP_LOCK:
-    if (check_global_access(thd, RELOAD_ACL))
+    if (check_global_access(thd, RELOAD_ACL, true))
+    {
+#ifndef NO_EMBEDDED_ACCESS_CHECKS
+      /*
+        In case there is no global privilege, check DB privilege for BACK LOCK.
+      */
+      if (first_table)
+      {
+        if (check_single_table_access(thd, LOCK_TABLES_ACL, first_table, true))
+        {
+          char command[30];
+          get_privilege_desc(command, sizeof(command), RELOAD_ACL|LOCK_TABLES_ACL);
+          my_error(ER_SPECIFIC_ACCESS_DENIED_ERROR, MYF(0), command);
+          goto error;
+        }
+      }
+      else // BACKUP UNLOCK
+      {
+        if (!thd->mdl_backup_lock)
```

Comment:
Suggested code comment:

We test mdl_backup_lock here because if a user could obtain a lock it would be silly to error and say you can't BACKUP UNLOCK (because its obvious you did a BACKUP LOCK). As BACKUP UNLOCK doesn't have a database reference, and there's no way we can check if the BACKUP LOCK privilege is missing. Testing thd->db would involve faking a TABLE_LIST structure which because of the depth of inspection in check_single_table_access make the faking likely to cause crashes or unintended effects. The outcome of this is if a user does an BACKUP UNLOCK without a BACKUP LOCKED table they may be a ER_SPECIFIC_ACCESS_DENIED error even though they do have the privilege. Its a bit different to what happens if the user has RELOAD_ACL where the error is silently ignored.